File: internal/repository/conference.go
package repository

import (
	"context"
	"errors"
	"fmt"

	"goshen/backend/internal/model"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

type ConferenceRepo struct{ db *pgxpool.Pool }

func NewConferenceRepo(db *pgxpool.Pool) *ConferenceRepo { return &ConferenceRepo{db: db} }

func (r *ConferenceRepo) ListUnpublishedSlugs(ctx context.Context) ([]string, error) {
	rows, err := r.db.Query(ctx, `SELECT slug FROM conference_pages WHERE is_published=false`)
	if err != nil {
		return nil, fmt.Errorf("list unpublished conference slugs: %w", err)
	}
	defer rows.Close()
	var slugs []string
	for rows.Next() {
		var s string
		if err := rows.Scan(&s); err != nil {
			return nil, err
		}
		slugs = append(slugs, s)
	}
	return slugs, rows.Err()
}

func (r *ConferenceRepo) Get(ctx context.Context, slug string) (*model.ConferenceAdminData, error) {
	var data model.ConferenceAdminData
	var pageID string

	err := r.db.QueryRow(ctx,
		`SELECT id::text, slug, label, is_published FROM conference_pages WHERE slug=$1`, slug,
	).Scan(&pageID, &data.Slug, &data.Label, &data.IsPublished)
	if err != nil {
		if errors.Is(err, pgx.ErrNoRows) {
			return nil, nil
		}
		return nil, fmt.Errorf("get conference page: %w", err)
	}
	data.ID = pageID

	_ = r.db.QueryRow(ctx,
		`SELECT hero_image_url, badge_text, headline, sub_text
		 FROM conference_hero WHERE page_id=$1`, pageID,
	).Scan(&data.Hero.HeroImageURL, &data.Hero.BadgeText, &data.Hero.Headline, &data.Hero.SubText)

	titleRows, err := r.db.Query(ctx,
		`SELECT section_key, title FROM conference_section_titles WHERE page_id=$1`, pageID)
	if err != nil {
		return nil, fmt.Errorf("get section titles: %w", err)
	}
	for titleRows.Next() {
		var key, title string
		if err := titleRows.Scan(&key, &title); err != nil {
			titleRows.Close()
			return nil, err
		}
		switch key {
		case "product_grid":
			data.Titles.ProductGrid = title
		case "workspace":
			data.Titles.Workspace = title
		case "solutions":
			data.Titles.Solutions = title
		case "contact":
			data.Titles.Contact = title
		}
	}
	titleRows.Close()

	_ = r.db.QueryRow(ctx,
		`SELECT description FROM conference_workspace WHERE page_id=$1`, pageID,
	).Scan(&data.WorkspaceDescription)

	solRows, err := r.db.Query(ctx,
		`SELECT id::text, room_size, title, description, kit_label,
		        COALESCE(image_url,''), COALESCE(image_url_2,''),
		        COALESCE(card1_name,''), COALESCE(card1_category,''), COALESCE(card1_sub_category,''),
		        COALESCE(card2_name,''), COALESCE(card2_category,''), COALESCE(card2_sub_category,''),
		        is_hidden, sort_order
		 FROM conference_room_solutions WHERE page_id=$1 ORDER BY sort_order ASC`, pageID)
	if err != nil {
		return nil, fmt.Errorf("get room solutions: %w", err)
	}
	var solutions []model.ConferenceSolutionData
	for solRows.Next() {
		var s model.ConferenceSolutionData
		if err := solRows.Scan(&s.ID, &s.RoomSize, &s.Title, &s.Description, &s.KitLabel,
			&s.ImageURL, &s.ImageURL2,
			&s.Card1Name, &s.Card1Category, &s.Card1SubCategory,
			&s.Card2Name, &s.Card2Category, &s.Card2SubCategory,
			&s.IsHidden, &s.SortOrder); err != nil {
			solRows.Close()
			return nil, err
		}
		solutions = append(solutions, s)
	}
	solRows.Close()

	for i := range solutions {
		itemRows, err := r.db.Query(ctx,
			`SELECT item FROM conference_room_kit_items WHERE room_solution_id=$1 ORDER BY sort_order ASC`,
			solutions[i].ID)
		if err != nil {
			return nil, fmt.Errorf("get kit items: %w", err)
		}
		var items []string
		for itemRows.Next() {
			var item string
			if err := itemRows.Scan(&item); err != nil {
				itemRows.Close()
				return nil, err
			}
			items = append(items, item)
		}
		itemRows.Close()
		if items == nil {
			items = []string{}
		}
		solutions[i].KitItems = items
	}
	if solutions == nil {
		solutions = []model.ConferenceSolutionData{}
	}
	data.Solutions = solutions

	prodRows, err := r.db.Query(ctx,
		`SELECT cp.id::text, p.id, p.name, p.category, COALESCE(p.sub_category,''), COALESCE(p.image_url,''), cp.is_hidden
		 FROM conference_products cp
		 JOIN products p ON p.id = cp.product_id
		 WHERE cp.page_id=$1 AND cp.section='product_grid' ORDER BY cp.sort_order ASC`, pageID)
	if err != nil {
		return nil, fmt.Errorf("get products: %w", err)
	}
	var products []model.ConferenceProductData
	for prodRows.Next() {
		var p model.ConferenceProductData
		if err := prodRows.Scan(&p.ID, &p.ProductID, &p.Name, &p.Category, &p.SubCategory, &p.ImageURL, &p.IsHidden); err != nil {
			prodRows.Close()
			return nil, err
		}
		products = append(products, p)
	}
	prodRows.Close()
	if products == nil {
		products = []model.ConferenceProductData{}
	}
	data.Products = products

	return &data, nil
}

func (r *ConferenceRepo) Save(ctx context.Context, data *model.ConferenceAdminData) error {
	tx, err := r.db.Begin(ctx)
	if err != nil {
		return fmt.Errorf("begin tx: %w", err)
	}
	defer tx.Rollback(ctx) //nolint:errcheck

	var pageID string
	err = tx.QueryRow(ctx,
		`INSERT INTO conference_pages (slug, label, is_published)
		 VALUES ($1, $2, $3)
		 ON CONFLICT (slug) DO UPDATE SET label=EXCLUDED.label, is_published=EXCLUDED.is_published
		 RETURNING id::text`,
		data.Slug, data.Label, data.IsPublished,
	).Scan(&pageID)
	if err != nil {
		return fmt.Errorf("upsert page: %w", err)
	}

	_, err = tx.Exec(ctx,
		`INSERT INTO conference_hero (page_id, hero_image_url, badge_text, headline, sub_text)
		 VALUES ($1, $2, $3, $4, $5)
		 ON CONFLICT (page_id) DO UPDATE SET
		   hero_image_url=EXCLUDED.hero_image_url, badge_text=EXCLUDED.badge_text,
		   headline=EXCLUDED.headline, sub_text=EXCLUDED.sub_text`,
		pageID, data.Hero.HeroImageURL, data.Hero.BadgeText, data.Hero.Headline, data.Hero.SubText,
	)
	if err != nil {
		return fmt.Errorf("upsert hero: %w", err)
	}

	type sectionRow struct{ key, title string }
	for _, s := range []sectionRow{
		{"product_grid", data.Titles.ProductGrid},
		{"workspace", data.Titles.Workspace},
		{"solutions", data.Titles.Solutions},
		{"contact", data.Titles.Contact},
	} {
		_, err = tx.Exec(ctx,
			`INSERT INTO conference_section_titles (page_id, section_key, title)
			 VALUES ($1, $2, $3)
			 ON CONFLICT (page_id, section_key) DO UPDATE SET title=EXCLUDED.title`,
			pageID, s.key, s.title,
		)
		if err != nil {
			return fmt.Errorf("upsert section title %s: %w", s.key, err)
		}
	}

	_, err = tx.Exec(ctx,
		`INSERT INTO conference_workspace (page_id, description)
		 VALUES ($1, $2)
		 ON CONFLICT (page_id) DO UPDATE SET description=EXCLUDED.description`,
		pageID, data.WorkspaceDescription,
	)
	if err != nil {
		return fmt.Errorf("upsert workspace: %w", err)
	}

	// Load existing room solutions so we can delete removed ones
	existingRows, err := tx.Query(ctx,
		`SELECT id::text, room_size FROM conference_room_solutions WHERE page_id=$1`, pageID)
	if err != nil {
		return fmt.Errorf("query room solutions: %w", err)
	}
	existing := map[string]string{} // roomSize -> id
	for existingRows.Next() {
		var id, rs string
		if err := existingRows.Scan(&id, &rs); err != nil {
			existingRows.Close()
			return err
		}
		existing[rs] = id
	}
	existingRows.Close()

	incoming := map[string]bool{}
	for i, s := range data.Solutions {
		incoming[s.RoomSize] = true
		var solID string
		imgURL := sqlNullStr(s.ImageURL)
		imgURL2 := sqlNullStr(s.ImageURL2)
		err = tx.QueryRow(ctx,
			`INSERT INTO conference_room_solutions
			   (page_id, room_size, title, description, kit_label, image_url, image_url_2,
			    card1_name, card1_category, card1_sub_category,
			    card2_name, card2_category, card2_sub_category,
			    is_hidden, sort_order)
			 VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15)
			 ON CONFLICT (page_id, room_size) DO UPDATE SET
			   title=EXCLUDED.title, description=EXCLUDED.description,
			   kit_label=EXCLUDED.kit_label, image_url=EXCLUDED.image_url,
			   image_url_2=EXCLUDED.image_url_2,
			   card1_name=EXCLUDED.card1_name, card1_category=EXCLUDED.card1_category,
			   card1_sub_category=EXCLUDED.card1_sub_category,
			   card2_name=EXCLUDED.card2_name, card2_category=EXCLUDED.card2_category,
			   card2_sub_category=EXCLUDED.card2_sub_category,
			   is_hidden=EXCLUDED.is_hidden, sort_order=EXCLUDED.sort_order
			 RETURNING id::text`,
			pageID, s.RoomSize, s.Title, s.Description, s.KitLabel,
			imgURL, imgURL2,
			s.Card1Name, s.Card1Category, s.Card1SubCategory,
			s.Card2Name, s.Card2Category, s.Card2SubCategory,
			s.IsHidden, i+1,
		).Scan(&solID)
		if err != nil {
			return fmt.Errorf("upsert room solution %q: %w", s.RoomSize, err)
		}

		_, err = tx.Exec(ctx,
			`DELETE FROM conference_room_kit_items WHERE room_solution_id=$1`, solID)
		if err != nil {
			return fmt.Errorf("clear kit items: %w", err)
		}
		for j, item := range s.KitItems {
			if item == "" {
				continue
			}
			_, err = tx.Exec(ctx,
				`INSERT INTO conference_room_kit_items (room_solution_id, item, sort_order)
				 VALUES ($1, $2, $3)`,
				solID, item, j+1,
			)
			if err != nil {
				return fmt.Errorf("insert kit item: %w", err)
			}
		}
	}

	for roomSize, id := range existing {
		if !incoming[roomSize] {
			_, err = tx.Exec(ctx, `DELETE FROM conference_room_solutions WHERE id=$1`, id)
			if err != nil {
				return fmt.Errorf("delete room solution: %w", err)
			}
		}
	}

	_, err = tx.Exec(ctx,
		`DELETE FROM conference_products WHERE page_id=$1 AND section='product_grid'`, pageID)
	if err != nil {
		return fmt.Errorf("delete products: %w", err)
	}
	for i, p := range data.Products {
		_, err = tx.Exec(ctx,
			`INSERT INTO conference_products (page_id, section, product_id, is_hidden, sort_order)
			 VALUES ($1,'product_grid',$2,$3,$4)`,
			pageID, p.ProductID, p.IsHidden, i+1,
		)
		if err != nil {
			return fmt.Errorf("insert product: %w", err)
		}
	}

	return tx.Commit(ctx)
}

func sqlNullStr(s string) any {
	if s == "" {
		return nil
	}
	return s
}

File: src/routes/conference.ts
// goshen-api/src/routes/conference.ts
import { Hono } from 'hono'
import { eq, asc } from 'drizzle-orm'
import { db } from '../db/client.js'
import {
  conferencePages, conferenceHero, conferenceSectionTitles,
  conferenceProducts, conferenceWorkspace, conferenceRoomSolutions,
  conferenceRoomKitItems, products,
} from '../db/schema.js'
import { ok, created, notFound, internalError, badRequest } from '../lib/response.js'
import { requireAuth } from '../middleware/auth.js'

const VALID_SLUGS = new Set(['enterprise', 'government', 'higher-education', 'hospitality'])

interface ConferencePutBody {
  isPublished?: boolean
  hero?: { heroImageUrl?: string; badgeText?: string; headline?: string; subText?: string }
  titles?: { productGrid?: string; workspace?: string; solutions?: string; contact?: string }
  workspaceDescription?: string
}

// Builds a full conference page object from DB
async function buildConferencePage(slug: string) {
  const [page] = await db.select().from(conferencePages).where(eq(conferencePages.slug, slug)).limit(1)
  if (!page) return null

  const [hero] = await db.select().from(conferenceHero).where(eq(conferenceHero.pageId, page.id)).limit(1)
  const titleRows = await db.select().from(conferenceSectionTitles)
    .where(eq(conferenceSectionTitles.pageId, page.id))
  const [workspace] = await db.select().from(conferenceWorkspace)
    .where(eq(conferenceWorkspace.pageId, page.id)).limit(1)

  const solutionRows = await db.select().from(conferenceRoomSolutions)
    .where(eq(conferenceRoomSolutions.pageId, page.id))
    .orderBy(asc(conferenceRoomSolutions.sortOrder))

  const solutionsWithKits = await Promise.all(solutionRows.map(async (s) => {
    const items = await db.select().from(conferenceRoomKitItems)
      .where(eq(conferenceRoomKitItems.roomSolutionId, s.id))
      .orderBy(asc(conferenceRoomKitItems.sortOrder))
    return { ...s, kitItems: items.map(i => i.item) }
  }))

  const productRows = await db
    .select({
      id: conferenceProducts.id,
      productId: conferenceProducts.productId,
      section: conferenceProducts.section,
      isHidden: conferenceProducts.isHidden,
      sortOrder: conferenceProducts.sortOrder,
      name: products.name,
      category: products.category,
      subCategory: products.subCategory,
      imageUrl: products.imageUrl,
    })
    .from(conferenceProducts)
    .innerJoin(products, eq(conferenceProducts.productId, products.id))
    .where(eq(conferenceProducts.pageId, page.id))
    .orderBy(asc(conferenceProducts.sortOrder))

  const titlesMap: Record<string, string> = {}
  for (const t of titleRows) titlesMap[t.sectionKey] = t.title

  return {
    id: page.id,
    slug: page.slug,
    label: page.label,
    isPublished: page.isPublished,
    hero: hero ? {
      heroImageUrl: hero.heroImageUrl,
      badgeText: hero.badgeText,
      headline: hero.headline,
      subText: hero.subText,
    } : null,
    titles: {
      productGrid: titlesMap['product_grid'] ?? '',
      workspace: titlesMap['workspace'] ?? '',
      solutions: titlesMap['solutions'] ?? '',
      contact: titlesMap['contact'] ?? '',
    },
    workspaceDescription: workspace?.description ?? '',
    solutions: solutionsWithKits,
    products: productRows,
  }
}

export const conferenceRoutes = new Hono()

// GET /api/v1/conference-pages/:slug (public)
conferenceRoutes.get('/conference-pages/:slug', async (c) => {
  try {
    const data = await buildConferencePage(c.req.param('slug'))
    if (!data) return notFound(c, 'conference page not found')
    return ok(c, data)
  } catch { return internalError(c) }
})

// GET /api/v1/admin/conference-pages/:slug (protected)
conferenceRoutes.get('/admin/conference-pages/:slug', requireAuth, async (c) => {
  try {
    const data = await buildConferencePage(c.req.param('slug'))
    if (!data) return notFound(c, 'conference page not found')
    return ok(c, data)
  } catch { return internalError(c) }
})

// PUT /api/v1/admin/conference-pages/:slug (protected)
// Accepts JSON body: { isPublished, hero, titles, workspaceDescription }
conferenceRoutes.put('/admin/conference-pages/:slug', requireAuth, async (c) => {
  try {
    const slug = c.req.param('slug')
    if (!VALID_SLUGS.has(slug)) return badRequest(c, 'invalid conference slug')

    const body = await c.req.json<ConferencePutBody>()

    const [page] = await db.select().from(conferencePages).where(eq(conferencePages.slug, slug)).limit(1)
    if (!page) return notFound(c, 'conference page not found')

    const hero = body.hero ?? {}
    await db.insert(conferenceHero).values({
      pageId: page.id,
      heroImageUrl: hero.heroImageUrl ?? '',
      badgeText: hero.badgeText ?? '',
      headline: hero.headline ?? '',
      subText: hero.subText ?? '',
    }).onConflictDoUpdate({
      target: conferenceHero.pageId,
      set: {
        heroImageUrl: hero.heroImageUrl ?? '',
        badgeText: hero.badgeText ?? '',
        headline: hero.headline ?? '',
        subText: hero.subText ?? '',
      },
    })

    // Section titles — map camelCase dashboard keys to snake_case section_key values
    const titles = body.titles ?? {}
    const titleEntries: [string, string][] = [
      ['product_grid', titles.productGrid ?? ''],
      ['workspace', titles.workspace ?? ''],
      ['solutions', titles.solutions ?? ''],
      ['contact', titles.contact ?? ''],
    ]
    for (const [sectionKey, title] of titleEntries) {
      await db.insert(conferenceSectionTitles)
        .values({ pageId: page.id, sectionKey, title })
        .onConflictDoUpdate({
          target: [conferenceSectionTitles.pageId, conferenceSectionTitles.sectionKey],
          set: { title },
        })
    }

    const workspaceDesc = body.workspaceDescription ?? ''
    await db.insert(conferenceWorkspace).values({ pageId: page.id, description: workspaceDesc })
      .onConflictDoUpdate({ target: conferenceWorkspace.pageId, set: { description: workspaceDesc } })

    if (body.isPublished !== undefined) {
      await db.update(conferencePages)
        .set({ isPublished: body.isPublished })
        .where(eq(conferencePages.id, page.id))
    }

    return ok(c, await buildConferencePage(slug))
  } catch {
    return internalError(c)
  }
})

// POST /admin/conference-pages/:slug/products
conferenceRoutes.post('/admin/conference-pages/:slug/products', requireAuth, async (c) => {
  try {
    const slug = c.req.param('slug')
    const [page] = await db.select().from(conferencePages).where(eq(conferencePages.slug, slug)).limit(1)
    if (!page) return notFound(c, 'conference page not found')
    const { productId, section } = await c.req.json<{ productId: number; section: string }>()
    if (!productId || !section) return badRequest(c, 'productId and section required')
    const count = await db.select().from(conferenceProducts)
      .where(eq(conferenceProducts.pageId, page.id)).then(r => r.length)
    const [row] = await db.insert(conferenceProducts)
      .values({ pageId: page.id, productId, section, sortOrder: count })
      .returning()
    return created(c, row)
  } catch (e: any) {
    if (e?.code === '23505') return c.json({ success: false, error: 'product already on this page' }, 409)
    return internalError(c)
  }
})

// PATCH /admin/conference-pages/:slug/products/:id
conferenceRoutes.patch('/admin/conference-pages/:slug/products/:id', requireAuth, async (c) => {
  try {
    const id = c.req.param('id')
    const { isHidden } = await c.req.json<{ isHidden: boolean }>()
    const [row] = await db.update(conferenceProducts).set({ isHidden }).where(eq(conferenceProducts.id, id)).returning()
    if (!row) return notFound(c, 'not found')
    return ok(c, row)
  } catch { return internalError(c) }
})

// DELETE /admin/conference-pages/:slug/products/:id
conferenceRoutes.delete('/admin/conference-pages/:slug/products/:id', requireAuth, async (c) => {
  try {
    const id = c.req.param('id')
    await db.delete(conferenceProducts).where(eq(conferenceProducts.id, id))
    return ok(c, null)
  } catch { return internalError(c) }
})

// POST /admin/conference-pages/:slug/solutions
conferenceRoutes.post('/admin/conference-pages/:slug/solutions', requireAuth, async (c) => {
  try {
    const slug = c.req.param('slug')
    const [page] = await db.select().from(conferencePages).where(eq(conferencePages.slug, slug)).limit(1)
    if (!page) return notFound(c, 'conference page not found')
    const body = await c.req.json<{
      roomSize: string; title: string; description: string; kitLabel: string; kitItems: string[]
      imageUrl: string; imageUrl2: string
      card1Name: string; card1Category: string; card1SubCategory: string
      card2Name: string; card2Category: string; card2SubCategory: string
      isHidden: boolean
    }>()
    const count = await db.select().from(conferenceRoomSolutions)
      .where(eq(conferenceRoomSolutions.pageId, page.id)).then(r => r.length)
    const [sol] = await db.insert(conferenceRoomSolutions).values({
      pageId: page.id,
      roomSize: body.roomSize,
      title: body.title ?? '',
      description: body.description ?? '',
      kitLabel: body.kitLabel ?? 'IMX ROOM KIT 30:',
      imageUrl: body.imageUrl ?? '',
      imageUrl2: body.imageUrl2 ?? '',
      card1Name: body.card1Name ?? '',
      card1Category: body.card1Category ?? '',
      card1SubCategory: body.card1SubCategory ?? '',
      card2Name: body.card2Name ?? '',
      card2Category: body.card2Category ?? '',
      card2SubCategory: body.card2SubCategory ?? '',
      isHidden: body.isHidden ?? false,
      sortOrder: count,
    }).returning()
    const items = (body.kitItems ?? []).filter(Boolean)
    for (let i = 0; i < items.length; i++) {
      await db.insert(conferenceRoomKitItems).values({ roomSolutionId: sol.id, item: items[i], sortOrder: i })
    }
    return created(c, sol)
  } catch (e: any) {
    if (e?.code === '23505') return c.json({ success: false, error: 'roomSize already exists for this page' }, 409)
    return internalError(c)
  }
})

// PUT /admin/conference-pages/:slug/solutions/:id
conferenceRoutes.put('/admin/conference-pages/:slug/solutions/:id', requireAuth, async (c) => {
  try {
    const id = c.req.param('id')
    const body = await c.req.json<{
      roomSize?: string; title?: string; description?: string; kitLabel?: string; kitItems?: string[]
      imageUrl?: string; imageUrl2?: string
      card1Name?: string; card1Category?: string; card1SubCategory?: string
      card2Name?: string; card2Category?: string; card2SubCategory?: string
      isHidden?: boolean
    }>()
    const { kitItems } = body
    const fields = {
      ...(body.roomSize !== undefined && { roomSize: body.roomSize }),
      ...(body.title !== undefined && { title: body.title }),
      ...(body.description !== undefined && { description: body.description }),
      ...(body.kitLabel !== undefined && { kitLabel: body.kitLabel }),
      ...(body.imageUrl !== undefined && { imageUrl: body.imageUrl }),
      ...(body.imageUrl2 !== undefined && { imageUrl2: body.imageUrl2 }),
      ...(body.card1Name !== undefined && { card1Name: body.card1Name }),
      ...(body.card1Category !== undefined && { card1Category: body.card1Category }),
      ...(body.card1SubCategory !== undefined && { card1SubCategory: body.card1SubCategory }),
      ...(body.card2Name !== undefined && { card2Name: body.card2Name }),
      ...(body.card2Category !== undefined && { card2Category: body.card2Category }),
      ...(body.card2SubCategory !== undefined && { card2SubCategory: body.card2SubCategory }),
      ...(body.isHidden !== undefined && { isHidden: body.isHidden }),
    }
    const [sol] = await db.update(conferenceRoomSolutions).set(fields).where(eq(conferenceRoomSolutions.id, id)).returning()
    if (!sol) return notFound(c, 'solution not found')
    if (kitItems !== undefined) {
      await db.delete(conferenceRoomKitItems).where(eq(conferenceRoomKitItems.roomSolutionId, id))
      const items = kitItems.filter(Boolean)
      for (let i = 0; i < items.length; i++) {
        await db.insert(conferenceRoomKitItems).values({ roomSolutionId: id, item: items[i], sortOrder: i })
      }
    }
    return ok(c, sol)
  } catch { return internalError(c) }
})

// DELETE /admin/conference-pages/:slug/solutions/:id
conferenceRoutes.delete('/admin/conference-pages/:slug/solutions/:id', requireAuth, async (c) => {
  try {
    const id = c.req.param('id')
    await db.delete(conferenceRoomSolutions).where(eq(conferenceRoomSolutions.id, id))
    return ok(c, null)
  } catch { return internalError(c) }
})
